Add a `RepeatingAnimationBuilder` API

bernaferrari opened the pull request:
Implements / Close https://github.com/flutter/flutter/issues/174011

<img width="1509" height="649" alt="diff" src="https://github.com/user-attachments/assets/cd961ddc-3b79-4ffd-8e9a-a806d5743f17" />

loic-sharma commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -117,7 +117,99 @@ class TweenAnimationBuilder<T extends Object?> extends ImplicitlyAnimatedWidget
+  ///  * [AnimationController.repeat], which this constructor replaces for simple use cases.
+  ///  * [TweenAnimationBuilder], the default constructor for single animations.
+  ///
+  /// ## Ownership
```
Could you move the Ownership section up above the code samples?

loic-sharma commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -117,7 +117,99 @@ class TweenAnimationBuilder<T extends Object?> extends ImplicitlyAnimatedWidget
+  ///  * [AnimationController.repeat], which this constructor replaces for simple use cases.
+  ///  * [TweenAnimationBuilder], the default constructor for single animations.
+  ///
+  /// ## Ownership
+  ///
+  /// The [TweenAnimationBuilder] takes full ownership of the provided [tween]
+  /// instance and mutates it. Once a [Tween] has been passed to a
+  /// [TweenAnimationBuilder], its properties should not be accessed or changed
+  /// anymore to avoid interference with the [TweenAnimationBuilder].
```
This content mirrors but is slightly different from the "Ownership of the Tween" content on TweenAnimationBuilder. I'd consider making a template out of the original content and just use the template here.

https://github.com/flutter/flutter/blob/6cbd98adc517ca7a20c724c4f0c5a5a385b95f74/packages/flutter/lib/src/widgets/tween_animation_builder.dart#L58C1-L67C34

bernaferrari replied:
the issue of making a template (now) where it is a separate component is that I can't repeat "The [TweenAnimationBuilder] takes full ownership" unless it accepts variable so might make sense to repeat now :( or  just omit or duplicate.

loic-sharma commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -117,7 +117,99 @@ class TweenAnimationBuilder<T extends Object?> extends ImplicitlyAnimatedWidget
+    bool paused = false,
+    required this.builder,
+    this.child,
+  }) : _isRepeating = true,
```
Oof.. This is not as nice as I had hoped. Maybe this would be better as a separate `RepeatedTweenAnimationBuilder` widget?

bernaferrari commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,168 @@ class _TweenAnimationBuilderState<T extends Object?>
+  ///
+  /// The builder receives the animation object and the optional child.
+  /// The current value can be accessed via animation.value.
+  final Widget Function(BuildContext context, Animation<T> animation, Widget? child) builder;
```
Good catch! Fixing

bernaferrari replied:
This is tricky. ValueWidgetBuilder<T> takes the actual value T, while  RepeatingTweenAnimationBuilder is currently taking Animation<T>.
The reason is, many examples require animation, so it doesn't make sense for us to do `animation.value` and then create a new FixedAnimation(value) or something like that, I thought it made more sense to just use animation directly instead of animation.value, but it makes it different than TweenAnimationBuilder, so it is tricky.

I wish Repeating had value instead of animation, but then it will make the code for example worse than right now. Any thoughts?

loic-sharma replied:
TLDR: We should use `ValueWidgetBuilder<T>` here as well.

Whether the builder accepts a `T` or a `Animation<T>` has implications on how many times the builder is called:

```dart
FooAnimationBuilder(
  tween: Tween<double>(begin: 0, end: 1),
  builder: (context, double sizeFactor, child) {
    // This builder is called each time sizeFactor changes.
    // The subtree is rebuilt on each tick.
    return ...;
  },
);

FooAnimationBuilder(
  tween: Tween<double>(begin: 0, end: 1),
  builder: (context, Animation<double> sizeFactorAnimation, child) {
    // This builder should be called ONCE. You or some child widget needs an
    // AnimatedBuilder to trigger a callback each time sizeFactorAnimation
    // changes. This lets you minimize the subtree that is rebuilt each
    // time the animation ticks.
    return ...;
  },
);
```

Counterintuitively, forcing the `FixedAnimation` is a good thing. It's a code smell that indicates we're using some API incorrectly:

* **Option 1**: Switch to a child widget that doesn't accept an `Animation<T>`. For the `SizeTransition` example, we should consider switching to a `ClipRect` + `Align` subtree instead.
* **Option 2**: Consider whether the child widget actually needs to accept an `Animation<T>`. If the child widget rebuilds itself on each tick, it should might as well just accept a `T`.
* **Option 3**: Don't use `RepeatingTweenAnimationBuilder`.

FWIW, I do think we should consider to offer a variant of `TweenAnimationBuilder` and `RepeatingTweenAnimationBuilder`, however this should likely be either a separate named constructor on `TweenAnimationBuilder`/`RepeatingTweenAnimationBuilder`, or, separate widgets altogether. I would punt this to a separate pull request though :)

loic-sharma commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,168 @@ class _TweenAnimationBuilderState<T extends Object?>
+    _controller = AnimationController(duration: widget.duration, vsync: this);
+    _updateAnimation();
+
+    // Start the animation if not paused and begin != end
```
Please add a period to the end of all comments :)

```suggestion
    // Start the animation if not paused and begin != end.
```

loic-sharma commented:
@bernaferrari Hey there, sorry for the long response time on my end. I think we should still merge this and use it only for exampples:

1. Let's make `RepeatingTweenAnimationBuilder`'s builder accept a `T` instead of `Animation<T>`. See [this comment](https://github.com/flutter/flutter/pull/174014#discussion_r2377296222) for details.
2. Use `RepeatingTweenAnimationBuilder` conservatively where it is clearly the right choice. I wouldn't update the progress indicators if that requires multiple rendering

This is still a good first step. We can continue to make `RepeatingTweenAnimationBuilder` more powerful (by expanding it to the `Animation<T>` pattern) later to use this in more areas :)

bernaferrari committed:
WIP

bernaferrari committed:
fix

loic-sharma commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,168 @@ class _TweenAnimationBuilderState<T extends Object?>
+    } else if (!wasAnimating ||
+        widget.paused != oldWidget.paused ||
+        widget.reverse != oldWidget.reverse) {
+      _controller.repeat(reverse: widget.reverse);
```
Hm a slight drawback here is that this will always goes forward, even if we paused while the animation was running in reverse. Unfortunately, this seems to be a limitation of the `AnimationController` API (we might need to add `AnimationController.pause()` and `AnimationController.resume()` methods to address this?). I think that's OK for now, we can always fix this later.

bernaferrari replied:
The AnimationController API is very VERY confusing, and I guess part of that went for our new Tween. Right now, "reverse" means: when it reaches the end it "reverses" until zero. Like 0->1->0, instead of 0->1, 0->1. It doesn't mean, like, "please revert the movement in the opposite direction"

At some point I considered cleaning/improving the AnimationController, but it would be breaking. Not sure if we should just (eventually) make InfiniteAnimationController or something like that, which Compose does (and is much cleaner).

bernaferrari committed:
simplify

loic-sharma commented on packages/flutter/test/widgets/tween_animation_builder_test.dart:
```diff
@@ -393,4 +393,195 @@ void main() {
+        ),
+      );
+
+      // Initial value should be 0.0
```
Nit: could you do a pass on this file's comment and add periods at the end of all comments?

loic-sharma reviewed (approved):
LGTM, nice work!

bernaferrari committed:
add period

dkwingsmt commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,163 @@ class _TweenAnimationBuilderState<T extends Object?>
+  /// Creates a widget that repeats a tween animation.
+  RepeatingTweenAnimationBuilder({
+    super.key,
+    required this.tween,
```
Do you think we can loosen the limitation of `tween` and instead allowing `Animatable` here? Reasoning:
* I expect that this class is mostly used to play looping animations. Currently a smooth loop needs either `reverse: true` or special periodic properties (such as angles). This class can not be used for, for example, a TranslateX that increases ease-in and then decreases ease-in, since the animation is asymmetrical and has the same beginning and end.
* By roughly examining the code I think the only place that use the special properties of `Tween` are using its `begin` and `end` in `shouldChange`.

bernaferrari committed:
Experiment

loic-sharma commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+    _updateAnimation();
+
+    // Start the animation if not paused and the animation spans different values.
+    if (!widget.paused && startValue != endValue) {
```
(GitHub seems to have lost my comment?)

Now that this accepts an `Animatable`, I can pass a `TweenSequence<Offset>` that goes right, then goes left back to where it started. We should run the animation even if the start and end values are the same.

bernaferrari commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+
+    // Update the animation if the curve or the animatable changed.
+    final bool animatableChanged =
+        !identical(widget.animatable, oldWidget.animatable) ||
```
Do you have any suggestion in how to compare them? Not sure == will work well either.

chunhtai replied:
runtimeType may not work correctly if they change tweens, the best way may be implementing equality in Animatable and subclass.

I am not sure why we didn't make animatable immutable

chunhtai replied:
discussed with the team offline, looks like the class is meant to be mutable, nothing we can do here I think. though we should convert to use == instead of identical in case some day we figure out a magic way to make it immutable and override == in the future.

dkwingsmt replied:
Can you explain what kind of animation abstraction you're talking about?

chunhtai replied:
that everything can be composed into a single animation<T>: `vsync -> apply curve ->drive animatable-> animation`

now that we use animatable as a function instead of a composable unit. which also strengthen that animatable should probably be immutable

loic-sharma replied:
To summarize this thread, we're in agreement with this solution:

> * Changing `curve` should not recreate the `CurvedAnimation` instance, since `CurvedAnimation.curve` is mutable.
> * We should not evaluate the final animation with `Animatable.evaluate`. Instead, we should manually evaluate the result every frame by calling `Animatable.transform` with the result of `CurvedAnimation`.
> 
> This way, neither new `curve` nor new `animatable` should lead to rebuilding `Animation` instances.
```

dkwingsmt commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+/// The animation continuously cycles from the [Animatable] evaluated at 0.0 to
+/// the value at 1.0.
+/// When [reverse] is true, the animation reverses direction when it reaches
+/// the end. When [paused] is true, the animation stops at its current value.
```
I suggest removing `pause` and add an optional `AnimationController`.
* We have constantly received PRs that add animation controllers to existing widgets. 
* The `pause` is replaceable by an animation controller.
* Controlling animation status with propagating widget property seems to go against the principle of performance optimization as described in `@macro flutter.widgets.TweenAnimationBuilder.optimizations`.
* IMO `pause` is an antipattern, because it's a imperative operation, which feels wrong when passed in as a widget property.

dkwingsmt replied:
> I disagree, but I suspect I don't fully understand your pushback - could expand on this? The animation controller is an imperative API. The RepeatingAnimationBuilder is a declarative API, that maps the user's intent down to the imperative API. In my mind, this nicely mirrors other patterns, like the declarative widget tree vs the imperative render tree.

Pardon me if I sounds like just echoing `@macro flutter.widgets.TweenAnimationBuilder.optimizations`, or if I'm understanding it wrong.

The animation pipeline is designed so that when you want to configure the animation (such as the begin and end of a tween) you don't have to pass the number all the way down through widget properties, which causes rebuilds. Instead you just drill in this change through mutable containers, such as `Tween`. The job of the animation pipeline is to pass in the correct mutable containers that contain configurations, not the correct configurations themselves.

In my opinion, the `pause` flag is similar to `begin` and `end`.

dkwingsmt replied:
I discussed with Loic offline. Since this class is meant for simpler occasions and should never have an animation controller, I agree that a `paused` is reasonable. Thanks for the discussion.

dkwingsmt commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+/// {@tool dartpad}
+/// This example shows a continuously rotating square that can be paused and resumed.
+///
+/// ** See code in examples/api/lib/widgets/tween_animation_builder/repeating_tween_animation_builder.0.dart **
```
Rename the example file (if we're renaming this class)

dkwingsmt commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+  /// In most cases this will be a [Tween] or [TweenSequence], but any
+  /// [Animatable] that produces values of type [T] is accepted.
+  ///
+  /// ## Ownership
```
Now that it's an animatable this widget doesn't seem to mutate it anymore, and we should no longer discourage storing it. This instance should now be managed by the parent (which might not needs special mentioning any more).

dkwingsmt commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+  /// The animatable to drive repeatedly.
+  ///
+  /// The animation continuously cycles from the value returned by evaluating
+  /// [animatable] at 0.0 to the value returned at 1.0.
```
I suggest removing this paragraph, since it's already mentioned in the class's doc.

dkwingsmt commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+/// Widget that animates an [Animatable] value and repeats indefinitely.
+///
+/// The animation continuously cycles from the [Animatable] evaluated at 0.0 to
+/// the value at 1.0.
```
I'd like to add a bit more detail to this paragraph. (The last sentence assumed the removal of `reverse`; feel free to tweak it otherwise.)
```dart
/// The animation continuously cycles from the value produced by the
/// [Animatable] at 0.0 to the value at 1.0. The [builder] receives the
/// current value produced by this [Animatable] and builds the child from it.
/// When the animation reaches 1.0 it jumps immediately back to the value at
/// 0.0.
```

dkwingsmt replied:
With `reverse` or `reverseMode` I've rephrased it as follows:
```dart
/// The animation continuously cycles from the value produced by the
/// [Animatable] at 0.0 to the value at 1.0. The [builder] receives the
/// current value produced by this [Animatable] and builds the child from it.
/// When the animation reaches 1.0, it either restarts from 0.0 or reverses
/// direction, based on the [repeatMode].
```

dkwingsmt commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+  /// The duration of the animation.
+  final Duration duration;
+
+  /// The curve to apply when animating the value.
```
I suggest a bit more details on how the curve affects the algorithm:
```dart
  /// The curve applied to the animation input before it is passed to the
  /// [animatable].
  ///
  /// In other words, the curve transforms the controller's 0.0..1.0 timeline
  /// and the resulting value is then fed into the [animatable].
```

dkwingsmt commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+
+  /// Called every time the animation value changes.
+  ///
+  /// The builder receives the current animation value and the optional child.
```
I suggest a different first paragraph to introduce this property more directly.
```suggestion
  /// A builder that creates the animated widget subtree.
  ///
  /// Invoked on every frame with the current animation value and the optional
  /// prebuilt child. The child can be incorporated to reuse a subtree that
  /// doesn't depend on the animation, avoiding unnecessary rebuilds.
```

loic-sharma replied:
Strictly speaking, this might not be called on every frame. For example, things like `Visibility` can mute a running animation.

Maybe:

```suggestion
  /// A builder that creates the animated widget subtree.
  ///
  /// The builder is called every time the animation value changes.
  ///
  /// The optional child can be used to avoid unnecessary rebuilds
  /// if a part of the subtree doesn't depend on the animation.
```

dkwingsmt commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+  /// The builder receives the current animation value and the optional child.
+  final ValueWidgetBuilder<T> builder;
+
+  /// The child widget to pass to the builder.
```
```suggestion
  /// An optional widget to pass to the builder.
```

loic-sharma commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+///
+/// The animation continuously cycles from the [Animatable] evaluated at 0.0 to
+/// the value at 1.0.
+/// When [reverse] is true, the animation reverses direction when it reaches
```
We do have [`ReverseTween`](https://api.flutter.dev/flutter/animation/ReverseTween-class.html) already. However, using `ReverseTween` would still be verbose:

```dart
RepeatingAnimationBuilder(
  animatable: TweenSequence<Offset>([
    TweenSequenceItem<Offset>(
      tween: Tween<Offset>(begin: Offset.zero, end: Offset(0, 100))
        .chain(CurveTween(curve: Curves.ease)),
      weight: 1.0,
    ),
    TweenSequenceItem<Offset>(
      tween: ReverseTween(
        Tween<Offset>(begin: Offset.zero, end: Offset(0, 100))
          .chain(CurveTween(curve: Curves.ease)),
      ),
      weight: 1.0,
    ),
  builder: ...,
),
```

I still strongly prefer just adding a `reverse` argument to `RepeatingAnimationBuilder`.

bernaferrari replied:
Not sure if I can help much on the discussion. On Android it would be:
```kt
val infiniteTransition = rememberInfiniteTransition(label = "position")

val offsetY by infiniteTransition.animateFloat(
    initialValue = 0f,
    targetValue = 100f,
    animationSpec = infiniteRepeatable(
        animation = tween(
            durationMillis = 1000,
            easing = FastOutSlowInEasing // equivalent to Curves.ease
        ),
        repeatMode = RepeatMode.Reverse
    ),
    label = "offsetY"
)

Box(
    modifier = Modifier.offset(y = offsetY.dp)
) {
    // Your content
}
```
or finite:
```kt
var trigger by remember { mutableStateOf(false) }

val offsetY by animateFloatAsState(
    targetValue = if (trigger) 100f else 0f,
    animationSpec = repeatable(
        iterations = 2, // 0→100→0 (one full cycle)
        animation = tween(1000),
        repeatMode = RepeatMode.Reverse
    ),
    label = "offsetY"
)
```

on SwiftUI it would be:
```swift
struct ContentView: View {
    @State private var isAnimating = false
    
    var body: some View {
        Text("Hello World")
            .offset(y: isAnimating ? 100 : 0)
            .animation(
                .easeInOut(duration: 1)
                    .repeatForever(autoreverses: true),
                value: isAnimating
            )
            .onAppear {
                isAnimating = true
            }
    }
}
```
or finite:
```swift
.animation(
    .easeInOut(duration: 1)
        .repeatCount(3, autoreverses: true), // Repeats 3 times
    value: isAnimating
)
```

dkwingsmt replied:
I talked with Loic offline and he misunderstood me for what I referred to for `ReversingAnimatable` (I referred to a `Animatable` that is basically a `TweenSequence` concatenating the target Animatable and the same Animatable reversed). Loic agreed that this class would be useful, but we struggled to come up with a good name.

We also agreed that an enum just like Android's `RepeatMode` would be good enough. Although it's technically the same as `bool reverse`, I think a repeat mode makes much more sense on this class semantically. With this option available, while a new `ReversingAnimatable` class would be awesome, it's no longer necessary and can wait for another PR.

What do you think of changing the boolean to a variable with a `RepeatMode` enum?

bernaferrari committed:
Improve based on comments

bernaferrari commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -212,3 +216,175 @@ class _TweenAnimationBuilderState<T extends Object?>
+///
+/// ** See code in examples/api/lib/widgets/tween_animation_builder/repeating_tween_animation_builder.0.dart **
+/// {@end-tool}
+///
```
<p>I pushed your PR comments fixed. While this PR should help a bit, I keep wondering how we can make Flutter animation as pleasant to write as Compose:</p>

<table>
  <tr>
    <td align="center">
      <img width="425" height="521" alt="image" src="https://github.com/user-attachments/assets/04b36952-a27a-4c49-88c5-ca93717ae01f" />
    </td>
    <td align="center">
      <img width="426" height="932" alt="image" src="https://github.com/user-attachments/assets/1c357fcc-8e48-4128-baf3-58dbdb50165a" />
    </td>
  </tr>
</table>

bernaferrari commented:
I think I addressed every comment. Thanks for all the suggestions. I think I rewrote this PR two or three times already. It is getting better every time.

bernaferrari committed:
Fix analyzer

bernaferrari commented:
I didn't re-request anything but I guess that's automated when I force pushed to latest. But feel free to review again if/when you want.

bernaferrari committed:
Fix import

Piinks commented:
Flyby comment from triage: it looks like there are a number of failing checks here.

bernaferrari commented:
Oh, I think tree was down. I'll rebase.

dkwingsmt commented on packages/flutter/lib/src/widgets/repeating_animation_builder.dart:
```diff
@@ -0,0 +1,167 @@
+  /// Each iteration starts over from the beginning once 1.0 is reached.
+  restart,
+
+  /// Each iteration runs forward, then reverses back to the beginning.
```
Can you document whether this means the `duration` is doubled for each actual iteration, or that the curve is calculated at double speed?

(Alternatively we can document it in `RepeatingAnimationBuilder` or even in `RepeatingAnimationBuilder.duration` but I think documenting it here is more useful if this enum is to be used in other widgets.)

dkwingsmt commented on packages/flutter/lib/src/widgets/repeating_animation_builder.dart:
```diff
@@ -0,0 +1,167 @@
+
+  /// The animatable to drive repeatedly. Typically this is a [Tween] or a
+  /// [TweenSequence], but any [Animatable] that produces values of type [T]
+  /// is accepted.
```
```suggestion
  /// The animatable to drive repeatedly.
  /// 
  /// Typically this is a [Tween] or a [TweenSequence], but any [Animatable]
  /// that produces values of type [T] is accepted.
```

dkwingsmt commented on packages/flutter/lib/src/widgets/tween_animation_builder.dart:
```diff
@@ -13,6 +13,7 @@ import 'package:flutter/animation.dart';
 
 import 'framework.dart';
 import 'implicit_animations.dart';
+import 'transitions.dart';
```
Why is this needed?

dkwingsmt reviewed (approved):
Comprehensive tests. Clear documents. Thank you so much!

loic-sharma reviewed (approved):
I left some minor documentation suggestions.

Thank you for doing this! This is truly outstanding work :)

bernaferrari commented:
I need to find the answer for this. Don't worry. I'll get that later today. Thanks for collaborating over 3 months for this.
<img width="429" height="183" alt="image" src="https://github.com/user-attachments/assets/9d002e59-79c7-4a6e-9145-dc4733722cbc" />

bernaferrari committed:
Fix PR
